Design Review for new DFW API

madhephaestus opened the pull request:
The users code is not isolated in seperate files and a pointer of DFW will be passed with the current time to the user call back lambda.

madhephaestus committed:
adding Kevin Harrington to the authors page

madhephaestus committed:
API for constructor changed

The Constructor takes the serial port number, but in the comments it
notes that Serial1 is the only valid one. THis has been changed to only
access Serial1 and removed the parameter.

madhephaestus committed:
begin refactoring

The competition code is being switched over to a function pointer model.

madhephaestus committed:
Partial new packet parser working

madhephaestus committed:
Switching back to original parser

madhephaestus committed:
fixing the start button detect

madhephaestus committed:
BAsic competition state machine working

madhephaestus committed:
DWF v2 is ready for design review

madhephaestus committed:
Adding DFW pointer to lambdas

The functions that the user defines, now include the passing of the DFW
object from the fundtion call to the user function. THe lets the user
functions to be in isolated files and avoids code clutter. It also
provides an example to the users of how to pass objects around between
sections of code.

AustinShalit commented on DFW.cpp:
```diff
@@ -136,172 +206,128 @@ int DFW::joysticklh(void) // returns joystick right vertical
 int DFW::start(void) // returns start button state 
```
All of the button functions should return bool (true meaning depressed, false otherwise).

madhephaestus replied:
i fixed this

AustinShalit commented on DFW.cpp:
```diff
@@ -136,172 +206,128 @@ int DFW::joysticklh(void) // returns joystick right vertical
 int DFW::start(void) // returns start button state 
 		{
 	int startB;
```
For all of these functions: Unused variable

madhephaestus replied:
i fixed this too

AustinShalit commented on DFW.h:
```diff
@@ -1,18 +1,33 @@
-// Serial controller library for DFWireless controllers
+/**
+ * Serial controller library for DFWireless controllers
```
We might have scope creep with the DFW library.  Now instead of just handling the controller, it is also handling all of the competition stuff.

madhephaestus replied:
This i disagree with. This library is not just a controller reader, it is the competition framework. Explicatly it is a framework, not a library. I talked this over with Miller, Stafford and Gennert. This model is what they want to see.

AustinShalit commented on DFW.h:
```diff
@@ -1,18 +1,33 @@
+ * For RBE Class use
+ * @author Joe St. Germain joest@wpi.edu
+ * @author Alex Taglieri (atags22)
+ * @author Kevin Harrington (madhephaestus)
```
I do not think we should include author tags.  Just use the git history.

madhephaestus replied:
Commit history is good too, except Joes contributions would be obliterated. I want to keep the author tags wherever possible.

AustinShalit commented on DFW.h:
```diff
@@ -1,18 +1,33 @@
-
+//
+//
+//
```
Remove these extra `//`

AustinShalit reviewed (changes requested):
This is probably a move in the right direction for the library.  A few general notes:
• See my note about scope creep.
• There is no style guide right now.  I recommend using either the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) or the [WPILib Style Guide](https://github.com/wpilibsuite/styleguide)
• This does not support the pause after autonomous to count up the score.  This is a very important feature.
• Please squash all commits

madhephaestus committed:
Updating API and packet parsing

The packet parsing is done in line rather than depending on the find
functions.

madhephaestus commented:
commits are squashed on the merge step in GitHub, btw...

madhephaestus committed:
Move the update function to a private method

madhephaestus committed:
fix the polarity of button presses

madhephaestus committed:
Add an additional state for waiting for teleop to start

The start button needs to be hit a second time to kick the framework
into teleop mode.
